lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -2,20 +2,597 @@
+        assert tool.tool_opts == "opt1;opt2"
+        assert tool.pbench_install_dir == mocked_install_dir
+        assert tool.logger == mocked_logger
+
+        with pytest.raises(NotImplementedError):
+            tool.install()
+        with pytest.raises(NotImplementedError):
+            tool.start(MockedPath())
+        with pytest.raises(NotImplementedError):
+            tool.stop()
+        with pytest.raises(NotImplementedError):
+            tool.wait()
+
+        with pytest.raises(ToolException) as excinfo:
+            Tool("who", "cares", MockedPath(exists=False), None)
+        assert (
+            str(excinfo.value)
+            == "pbench installation directory does not exist: /foo/bar"
+        )
+
+    @staticmethod
+    def test_create_process_with_logger(monkeypatch):
+        # First we need a mock of the subprocess.Popen() constructor, which
+        # will just record what the constructor was called with.
+        class MockPopen:
+            def __init__(self, args, cwd=None, stdin=None, stdout=None, stderr=None):
+                self.args = args
+                self.cwd = cwd
+                self.stdin = stdin
+                self.stdout = stdout
+                self.stderr = stderr
+
+        monkeypatch.setattr("subprocess.Popen", MockPopen)
+
+        # Next we need a mock of the threading.Thread() constructor, which
+        # will also just record what it was called with, and also offer the
+        # "daemon" field, and a "start()" method to be called.
+        the_thread = None
+
+        class MockThread:
+            def __init__(self, target: Any = None, args: Any = None):
+                self.target = target
+                self.args = args
+                self.daemon = False
+                self.start_called = False
+                nonlocal the_thread
+                the_thread = self
+
+            def start(self):
+                self.start_called = True
+
+        monkeypatch.setattr("threading.Thread", MockThread)
+
+        # Finally we need mocked out arguments to the Tool() constructor so
+        # that we can check the mocks above were used as expected.
+        class MockedLogger:
+            def getChild(self, arg: Any) -> str:
+                return f"parent.child({arg})"
+
+        mocked_install_dir = MockedPath()
+        mocked_logger = MockedLogger()
+        tool = Tool("test-tool", "opt1;opt2", mocked_install_dir, mocked_logger)
+
+        # Now we can invoke the method under test with some fake arguments to
+        # verify it does the right thing.
+        args = ["command", "arg1", "arg2"]
+        cwd = MockedPath()
+        process = tool._create_process_with_logger(args, cwd)
+
+        assert the_thread is not None
+        assert the_thread.target is log_raw_io_output
+        assert the_thread.args == (subprocess.PIPE, "parent.child(logger)")
+        assert the_thread.daemon is True
+        assert the_thread.start_called is True
```

Comment:
When in that mocked Thread object can we execute the asserts?

When the `MockThread.start()` method is called?

Isn't it the case that we are trying to verify that the `.start()` method is called?  So if it is not called, how will we know?

Reply:
I sometimes endeavor to resist the temptation of creating a conversation to make a joke (though I'm not always successful), but since this is Webb's conversation ...

`<lame joke follows>`
You should just code the mocked `.start` to assert that it's called, of course!
`<end lame joke>`

No, I agree in this case that the mock object needs to track the calls so you can assert externally. If we had C++ then we might make that assert in the mock's destructor at the end of the test method, but in Python that'd be unreliable.

Reply:
> if it is not called, how will we know?

That's a fair point; but that could be done with a simple flag (or a counter), rather than having to have a whole structure.

> When in that mocked Thread object can we execute the asserts?
> 
> When the MockThread.start() method is called?

Yes, we should be able to, since they are executed in the context of the "main thread" of the test.  But, it's possible that the test failure detection depends on exceptions propagating up to the outer PyTest frame, and it's possible that something in the CUT could interfere with the propagation of the `AssertError` exception, so, yeah, if we're going to be fully paranoid, it's probably best not to have `assert` statements in mock functions.  😝  (I just prefer having the checks closer to the stuff that they are supposed to check, and doing the capture requires a bunch of infrastructure which I would be happier if we could do without.)